### ISSUE TITLE
RGBA Image issues

### ISSUE THREAD

**Issue by wyhinton** [QuantumBadger/Speedy2D#19]
I'm running into to some unexpected behavior when trying to draw images with an alpha channel:

OS: Windows 10
```
speedy2d = { version = "1.0.7", git = "https://github.com/QuantumBadger/Speedy2D", default-features = true }
image = "0.23.12"
```

source image:
![image](https://user-images.githubusercontent.com/38958118/113206581-a000cf00-923d-11eb-97fc-260ddabb26f3.png)

result:
![image](https://user-images.githubusercontent.com/38958118/113206496-8790b480-923d-11eb-9bad-b939f2b432e2.png)

Note that this png is opaque. 
However trying to create an image that definitely has an alpha channel, I get an app crash with (exit code: 0xc000041d)

![rust-logo](https://user-images.githubusercontent.com/38958118/113206919-0a197400-923e-11eb-9fa0-6aea7676c1be.png)

```lang-rust
fn has_alpha(img: DynamicImage) -> bool{
    img.color().has_alpha()
}
```
I created this helper function to determine if an image has an alpha. While the smiey image it returns false, with the rust logo it returns true. I then match that bool to ```speedy2d::image::ImageDataType```. 

So in summary, when passed a PNG that does not have an alpha channel, the image is not being drawn correctly. When passed a png with an alpha channel, I'm getting an unhandled crash. 

I created a repo demonstrating these issues [here](https://github.com/wyhinton/speedy_2d_issue)

**Comment by QuantumBadger** [comment]
Thanks for the report!

I've had a look at the repo, and it looks like the issue is that the image bytes are being provided in the wrong format. The bytes need to be converted to either RGB or RGBA format before being passed in.

If you use the following code to load the image, then it should work fine:

```rust
        let img: DynamicImage = ImageReader::open("rust.png").unwrap().decode().unwrap();
        let has_alpha = has_alpha(&img);
        let (ci_width, ci_height) = img.dimensions();
        let ci_bytes = match has_alpha {
            true => img.into_rgba8().into_raw(),
            false => img.into_rgb8().into_raw()
        };
        let img_type = match has_alpha {
            true => speedy2d::image::ImageDataType::RGBA,
            false => speedy2d::image::ImageDataType::RGB,
        };
```

The calls to `into_rgba8()` and `into_rgb8()` convert the image to the correct type.

To improve the error logging and avoid a crash, I've added a check for the number of bytes being passed in:

https://github.com/QuantumBadger/Speedy2D/commit/fc57b8dac23196b09862d6719fc360ab2ba741ea

If you run `cargo update` then this change should be included in your local build, and the original code should produce a sensible error message!

If this doesn't resolve your issue (or you have any other questions) please let me know!

**Comment by wyhinton** [comment]
Hey, thanks for your response. 

Just wondering, did you get both of the images I included in the repo to draw correctly?

I'm still experiencing some issues.

![image](https://user-images.githubusercontent.com/38958118/113241624-0f49e380-927d-11eb-86f1-19316d7195da.png)

**Comment by QuantumBadger** [comment]
@wyhinton Yes I tried both images (the face and the Rust logo) and they both worked correctly with the updated code.

If you'd like I'm happy to take a look at the code used to draw the image in your last comment, if you upload it to the repo.

**Comment by wyhinton** [comment]
Sure, I updated the [repo ](https://github.com/wyhinton/speedy_2d_issue/tree/main/speedy_2d_png_issue/imgs)with the code. 

The repo just uses the smiley but I'm still getting the original issue with drawing the smiley (after making sure to run cargo update):
![image](https://user-images.githubusercontent.com/38958118/113383564-34f4ec80-9352-11eb-8195-50b9604707c9.png)

**Comment by QuantumBadger** [comment]
Thanks, I'll re-open this for now as I can reproduce the issue.

When I downloaded the image before, it looks like GitHub re-encoded it and added an alpha channel. The original image in the repo seems to use an 8-bit colour map and no alpha channel.

I exported the bytes from `into_raw()` to a file and GIMP was able to open it just fine, so it looks like the decoded image data is correct. I've also added a test for loading an image in RGB mode, which passes correctly. So I'm not sure what the issue is yet.

**Comment by QuantumBadger** [comment]
I added code to dump the raw pixels right after `glTexImage2D` is called, and the raw data is correct and can be opened with GIMP.

APITrace says that the image is being uploaded to the GPU as follows:

```
glTexImage2D(
    target = GL_TEXTURE_2D,
    level = 0,
    internalformat = GL_RGB8,
    width = 225,
    height = 224,
    border = 0,
    format = GL_RGB,
    type = GL_UNSIGNED_BYTE,
    pixels = [binary data, size = 147.874 kb])
```

Which all looks fine to me. However in the APITrace texture preview, it appears incorrect as in the image from @wyhinton.

Changing the way the image is loaded from `RGB` to `RGBA` causes it to render correctly.

This is very mysterious and it's possible that I'm missing something obvious :) It's also possible that there's some weird quirk in OpenGL that is causing this. Obviously this is an NPOW2 texture with an odd-numbered width, but in theory that shouldn't cause an issue here.

**Comment by QuantumBadger** [comment]
It did in fact turn out to be a weird OpenGL quirk :) It assumes the vertical byte stride will be a multiple of 4 unless you specify otherwise.

This should now be fixed! @wyhinton would you mind doing a `cargo update` please and checking that all issues are resolved for you?

**Comment by wyhinton** [comment]
Both the logo and smiley are drawing correctly now for me, thanks!
